core/fetcher.go

```diff
@@ -161,73 +172,81 @@ func (f *BlockFetcher) ValidatorSet(ctx context.Context, height int64) (*types.V
-	if err != nil {
-		close(f.doneCh)
-		f.isListeningForBlocks.Store(false)
-		return nil, err
-	}
+func (f *BlockFetcher) runSubscriber() (chan types.EventDataSignedBlock, error) {
+	signedBlockCh := make(chan types.EventDataSignedBlock, 1)
 
-	log.Debug("created a subscription. Start listening for new blocks...")
-	signedBlockCh := make(chan types.EventDataSignedBlock)
 	go func() {
 		defer close(f.doneCh)
 		defer close(signedBlockCh)
-		defer func() { f.isListeningForBlocks.Store(false) }()
 		for {
 			select {
-			case <-ctx.Done():
+			case <-f.ctx.Done():
+				return
+			default:
+			}
+
+			subscription, err := f.client.SubscribeNewHeights(f.ctx, &coregrpc.SubscribeNewHeightsRequest{})
+			switch {
+			case err == nil:
+				log.Debug("fetcher: subscription created")
+			case errors.Is(err, context.Canceled):
 				return
 			default:
-				resp, err := subscription.Recv()
-				if err != nil {
-					log.Errorw("fetcher: error receiving new height", "err", err.Error())
-					_, ok := status.FromError(err) // parsing the gRPC error
-					if ok {
-						// ok means that err contains a gRPC status error.
-						// move on another round of resubscribing.
-						return
-					}
-					continue
-				}
-				withTimeout, ctxCancel := context.WithTimeout(ctx, 10*time.Second)
-				signedBlock, err := f.GetSignedBlock(withTimeout, resp.Height)
-				ctxCancel()
-				if err != nil {
-					log.Errorw("fetcher: error receiving signed block", "height", resp.Height, "err", err.Error())
-					// sleeping a bit to avoid retrying instantly and give time for the gRPC connection
-					// to recover automatically.
-					time.Sleep(time.Second)
-					continue
-				}
-				select {
-				case signedBlockCh <- types.EventDataSignedBlock{
-					Header:       *signedBlock.Header,
-					Commit:       *signedBlock.Commit,
-					ValidatorSet: *signedBlock.ValidatorSet,
-					Data:         *signedBlock.Data,
-				}:
-				case <-ctx.Done():
-					return
-				}
+				// try re-subscribe in case of any errors that can come during subscription. gRPC
+				// retry mechanism has a back off on retries, so we don't need timers anymore.
+				log.Errorw("fetcher: failed to subscribe to new block events", "err", err)
+				continue
+			}
+
+			err = f.receive(f.ctx, signedBlockCh, subscription)
+			if err != nil {
+				log.Errorw("fetcher: error receiving new height", "err", err.Error())
+				continue
 			}
 		}
 	}()
-
 	return signedBlockCh, nil
 }
 
+func (f *BlockFetcher) receive(
+	ctx context.Context,
+	signedBlockCh chan types.EventDataSignedBlock,
+	subscription coregrpc.BlockAPI_SubscribeNewHeightsClient,
+) error {
+	log.Error("fetcher: started listening for new blocks")
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		resp, err := subscription.Recv()
+		if err != nil {
+			return err
+		}
+
+		withTimeout, ctxCancel := context.WithTimeout(ctx, 10*time.Second)
```

Comment:
It is nice to have a timeout here, but the value of 10 seconds seems to be low. It might fail due to some networking temporary problems. Would also be nice to make it configurable.

Reply:
Added a TODO